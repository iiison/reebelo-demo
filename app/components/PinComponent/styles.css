@import "../../styles/base/_variables";

.pinComponent{
  margin: 5px;
  /* padding: 0 20px; */
}

.button {
  height: 50px;
  width: 292px;
  color: $white;
  background-color: rgb(0, 39, 106);
  border-radius: 5px;
  font-size: $M;
  margin: 5px;
  cursor: pointer;
}
.button:disabled{
    opacity: 0.3;
}

.heading1 {
  font-size: $M;
  font-weight: $font-bold;
  margin-top: 3%;
}

.heading2 {
  font-size: $M;
  font-weight: $font-normal;
  margin-top: 5px;
  margin-bottom: 10px;
  padding: 0 40px;
}

.helperText {
  font-size: $S;
}

.error {
    font-size: $S;
    color: $negative
}

.inputComponent {
  padding: 10px;

  &.pinError {
    input {
      border-color: $negative;
      color: $negative;
    }
  }
}

.skipFlow {
  font-size: $S;
  margin-top: 20px;
}

.visibilityToggle img {
    width: 20px;
    margin-left: 10px;
}
