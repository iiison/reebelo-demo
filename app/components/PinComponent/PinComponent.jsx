import React, { useState, useRef, useEffect } from 'react'
import { useSelector } from 'react-redux'
import { PropTypes } from 'prop-types'
import { useDispatch } from 'react-redux'

import {
  Loader,
  PinInputUnit,
  Timer,
} from '$COMPONENTS'
import errorTypeDetailMap from '$CONFIGS/errorTypeDetailMap'
import { ShowIcon, HideIcon } from '$COMPONENTS/icons/index';
import { setPinFailure } from '$RMODULES/pin/pin'
import { useSetState } from '$UTILS/reactUtils'

import styles from './styles.css'

const handleInputChange = ({ setPin, setValidationError, dispatch, error }) => (index) => (event) => {
  event.persist()
  const val = event.currentTarget.value

  if (error) {
    dispatch(setPinFailure(''))
    setValidationError('')
  }

  setPin({ [index]: val })
}

const handleKeyUp = (setFocus) => (index) => (event) => {
  event.persist()

  const { keyCode } = event

  if (keyCode && keyCode >= 48 && keyCode <= 57 && index !== 3) {
    setFocus(index + 1)
  }

  if (keyCode && keyCode === 8 && index !== 0) {
    setFocus(index - 1)
  }
}

const handlePinToggle = (type, setType) => (event) => {
  event.preventDefault();

  if (type === 'tel')
    setType('password');
  else
    setType('tel')
}

const handleSubmit = ({ tries, onSubmitProp, pinState, setValidationError }) => () => {
  if (!pinState) {
    setValidationError('Pin is invalid.')

    return
  }

  const pin = pinState.join('')

  if (pin && pin.length === 4) {
    setValidationError('')
    onSubmitProp(pin, tries)

    return
  }
  setValidationError('Pin is invalid.')
}

const chooseLoaderMessage = ({ loadingMessage, timeout,  }) => {
  if (timeout === 0) {
    return 'Operation timed out.'
  }

  return loadingMessage
}

const PinComponent = ({
  error,
  tries,
  heading,
  onSubmit,
  skipText,
  subHeading,
  helperText,
  isFetching,
  isVerified,
  loadingMessage,
  onSkip : onSkipProp
}) => {
  const [pin, setPin] = useSetState()
  const [timeout, setUserTimeout] = useState()
  const [type, setType] = useState('password')
  const [validationError, setValidationError] = useState('')
  const [focusedInput, setFocusedInput] = useState(0)

  const dispatch = useDispatch()

  const onSkip = (reason) => {
    setUserTimeout(0)
    onSkipProp(reason)
  };

  let configs = useSelector(({ configs }) => configs)
  let errorMessage

  if (error) {
    errorMessage = error.includes(' ') ? error : errorTypeDetailMap[error] || errorTypeDetailMap.generic
  }

  const pageErrors = validationError || errorMessage
  const handleInputChangeWithState = handleInputChange({ setPin, setValidationError, dispatch, error : pageErrors })
  const handlePinToggleWithState = handlePinToggle(type, setType)
  const handleKeyUpWithState = handleKeyUp(setFocusedInput)
  const handleFormSubmit = handleSubmit({
    tries,
    onSubmitProp : onSubmit,
    pinState : Object.values(pin),
    setValidationError
  })
  const inputRefs = [
    useRef(null),
    useRef(null),
    useRef(null),
    useRef(null),
  ]

  useEffect(() => {
    if (focusedInput !== undefined && inputRefs[focusedInput].current) {
      inputRefs[focusedInput].current.focus()
    }
  }, [focusedInput])

  return (
    <div className={`${styles.pinComponent}`}>
      <div className={`grid col-12 t-center`}>
        <h1 className={`col-12 t-capitalize ${styles.heading1}`}>{heading}</h1>
        <h2 className={`col-12 t-capitalize ${styles.heading2}`}>{subHeading}</h2>
        <div className={`col-12 grid-center grid-middle ${styles.inputComponent} ${error && styles.pinError}`}>
          {
            Array(4).fill().map((val, index) => {
              const props = {
                refVal : inputRefs[index],
                value : pin[index],
                type : type,
                events : {
                  onChange : handleInputChangeWithState(index),
                  onKeyUp : handleKeyUpWithState(index)
                }
              }

              return <PinInputUnit key={index} {...props} />
            })
          }
          <a
            href='#'
            className={`${styles.visibilityToggle}`}
            onClick={handlePinToggleWithState}
          >
            <img src={type === 'tel' ? HideIcon : ShowIcon} />
          </a>
        </div>
        {pageErrors ?
          <p className={`col-12 ${styles.error}`}>{pageErrors}</p> :
          <p className={`col-12 ${styles.helperText}`}>{helperText}</p>
        }

        <div className={`col-12 grid-center`}>
          <button type='button'
            disabled={Object.values(pin).join('').length === 4 ? false : true || isFetching}
            className={`${styles.button} margin-top-s`}
            onClick={handleFormSubmit}>
            {isFetching ? 'VERIFYING' : 'CONFIRM'}
          </button>
        </div>

        {configs.pinView === 'VERIFY' && <Timer seconds={configs.webResourceThresholdTime * 60} onSkip={onSkip} />}

        <a href='#' onClick={onSkip} className={`col-12 ${styles.skipFlow}`}>{skipText}</a>
        {
          (isFetching === true || isVerified === true) && 
            <Loader
              message={chooseLoaderMessage({ loadingMessage, timeout })}
              success={isVerified}
            />
        }
      </div>
    </div>
  )
}

PinComponent.propTypes = {
  heading        : PropTypes.string.isRequired,
  subHeading     : PropTypes.string.isRequired,
  onSubmit       : PropTypes.func.isRequired,
  onSkip         : PropTypes.func.isRequired,
  error          : PropTypes.string,
  helperText     : PropTypes.string,
  skipText       : PropTypes.string,
  isFetching     : PropTypes.bool,
  loadingMessage : PropTypes.string,
  timer          : PropTypes.string,
  tries          : PropTypes.number,
  isVerified     : PropTypes.bool
}

export default PinComponent
export {
  handleInputChange,
  handleSubmit,
  handleKeyUp
}

