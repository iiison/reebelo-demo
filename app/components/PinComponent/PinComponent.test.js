import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

import PinComponent, {
  handleInputChange,
  handleSubmit,
  handleKeyUp
} from './PinComponent'

const defaultProps = {
  heading    : 'Heading',
  subHeading : 'Subheading',
  onSubmit   : jest.fn(),
  error      : ''
}

describe('>> COMPONENT -- Tests PinComponent ', () => {
  describe('• handleInputChange function tests', () => {
    it('triggers the handleInputChange function', () => {
      const index = 0
      const setPinMock = jest.fn()
      const eventMock = {
        persist       : jest.fn(),
        currentTarget : {
          value : 'abc'
        }
      }

      handleInputChange(setPinMock)(index)(eventMock)

      expect(eventMock.persist).toHaveBeenCalled()
      expect(setPinMock).toHaveBeenCalledWith({
        [index] : eventMock.currentTarget.value
      })
    })
  })

  describe('• handleKeyUp function tests', () => {
    let setFocusMock, eventMock

    beforeEach(() => {
      setFocusMock = jest.fn()
      eventMock = {
        persist       : jest.fn(),
        currentTarget : {
          value : 1
        }
      }
    })

    it('should call setFocus function with 1', () => {
      handleKeyUp(setFocusMock)(0)({
        ...eventMock,
        keyCode : 48
      })

      expect(setFocusMock).toHaveBeenCalledWith(1)
    })

    it('should not call setFocus function due to the last input', () => {
      handleKeyUp(setFocusMock)(3)({
        ...eventMock,
        keyCode : 48
      })

      expect(setFocusMock).not.toHaveBeenCalled()
    })

    it('should not call setFocus function due to wrong keyCode', () => {
      handleKeyUp(setFocusMock)(0)({
        ...eventMock,
        keyCode : 200
      })

      expect(setFocusMock).not.toHaveBeenCalled()
    })

    it('tests backspace key press to set focus on previous input', () => {
      handleKeyUp(setFocusMock)(1)({
        ...eventMock,

        keyCode : 8
      })

      expect(setFocusMock).toHaveBeenCalledWith(0)
    })

    it('tests backspace key press to keep focus on same input', () => {
      handleKeyUp(setFocusMock)(0)({
        ...eventMock,

        keyCode : 8
      })

      expect(setFocusMock).not.toHaveBeenCalled()
    })
  })

  describe('• handleSubmit function tests', () => {
    it('tests handleSubmit happy flow', () => {
      const submitPropMock = jest.fn()
      const pinState = [1, 3, 2, 4]
      const setValidationErrorMock = jest.fn()

      handleSubmit({
        pinState,
        onSubmitProp       : submitPropMock,
        setValidationError : setValidationErrorMock
      })()

      expect(setValidationErrorMock).toHaveBeenCalledWith('')
      expect(submitPropMock).toHaveBeenCalledWith(pinState.join(''))
    })

    it('tests handleSubmit with empty pin', () => {
      const submitPropMock = jest.fn()
      const pinState = []
      const setValidationErrorMock = jest.fn()

      handleSubmit({
        pinState,
        onSubmitProp       : submitPropMock,
        setValidationError : setValidationErrorMock
      })()

      expect(setValidationErrorMock).toHaveBeenCalledWith('Pin is invalid.')
      expect(submitPropMock).toBeCalledTimes(0)
    })

    it('tests handleSubmit with undefined pin', () => {
      const submitPropMock = jest.fn()
      const pinState = undefined
      const setValidationErrorMock = jest.fn()

      handleSubmit({
        pinState,
        onSubmitProp       : submitPropMock,
        setValidationError : setValidationErrorMock
      })()

      expect(setValidationErrorMock).toHaveBeenCalledWith('Pin is invalid.')
      expect(submitPropMock).toBeCalledTimes(0)
    })

    it('tests handleSubmit with partial pin', () => {
      const submitPropMock = jest.fn()
      const pinState = [1, 3]
      const setValidationErrorMock = jest.fn()

      handleSubmit({
        pinState,
        onSubmitProp       : submitPropMock,
        setValidationError : setValidationErrorMock
      })()

      expect(setValidationErrorMock).toHaveBeenCalledWith('Pin is invalid.')
      expect(submitPropMock).toBeCalledTimes(0)
    })
  })

  describe('• PinComponent DOM Tests', () => {
    it('PinComponent Snapshot Test', () => {
      const props = defaultProps
      const component = renderer
        .create(<PinComponent {...props} />)
        .toJSON()

      expect(component).toMatchSnapshot();
    })

    it('Checks if input count is 4', () => {
      const props = defaultProps
      const component = shallow(<PinComponent {...props} />)

      const inputCount = component.find('PinInputUnit').length
      const expectedCount = 4

      expect(inputCount).toBe(expectedCount)
    })

    it('Triggers the Form submission with invalid values', () => {
      const props = defaultProps
      const component = mount(<PinComponent {...props} />)

      component.find('button').simulate('click')

      expect(props.onSubmit).toBeCalledTimes(0)
    })
  })
})

