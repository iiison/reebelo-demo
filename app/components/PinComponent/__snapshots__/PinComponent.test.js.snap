// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>> COMPONENT -- Tests PinComponent  â€¢ PinComponent DOM Tests PinComponent Snapshot Test 1`] = `
<div
  className="pinComponent"
>
  <div
    className="grid col-12 t-center"
  >
    <h1
      className="col-12 heading1"
    >
      Heading
    </h1>
    <h2
      className="col-12 heading2"
    >
      Subheading
    </h2>
    <div
      className="col-12-center inputComponent"
    >
      <input
        className="t-center pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="password"
        value=""
      />
      <input
        className="t-center pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="password"
        value=""
      />
      <input
        className="t-center pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="password"
        value=""
      />
      <input
        className="t-center pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="password"
        value=""
      />
      <a
        className="visibilityToggle"
        href="#"
        onClick={[Function]}
      >
        <img
          src={Object {}}
        />
      </a>
    </div>
    <p
      className="col-12 helperText"
    />
    <p
      className="col-12"
    >
      
    </p>
    <div
      className="col-12-center"
    >
      <button
        className="button"
        onClick={[Function]}
        type="button"
      >
        CONFIRM
      </button>
    </div>
    <a
      className="col-12 skipFlow"
      href="#"
    />
  </div>
</div>
`;
